crates/sui-package-resolver/src/store.rs

```diff
@@ -0,0 +1,225 @@
+    /// Latest version of the object at `id`.
+    async fn version(&self, id: AccountAddress) -> Result<SequenceNumber>;
+
+    /// Read package contents. Fails if `id` is not an object, not a package, or is malformed in
+    /// some way.
+    async fn fetch(&self, id: AccountAddress) -> Result<Package>;
+
+    /// Store package `object` in the underlying store
+    async fn update(&self, object: &Object) -> Result<()>;
+}
+
+pub struct DbPackageStore(pub(crate) IndexerReader);
+
+#[async_trait]
+impl PackageStore for DbPackageStore {
+    async fn version(&self, id: AccountAddress) -> Result<SequenceNumber> {
+        let query = objects::dsl::objects
+            .select(objects::dsl::object_version)
+            .filter(objects::dsl::object_id.eq(id.to_vec()));
+
+        let Some(version) = self
+            .0
+            .run_query_async(move |conn| query.get_result::<i64>(conn).optional())
+            .await?
+        else {
+            return Err(Error::PackageNotFound(id));
+        };
+
+        Ok(SequenceNumber::from_u64(version as u64))
+    }
+
+    async fn fetch(&self, id: AccountAddress) -> Result<Package> {
+        let query = objects::dsl::objects
+            .select((
+                objects::dsl::object_version,
+                objects::dsl::serialized_object,
+            ))
+            .filter(objects::dsl::object_id.eq(id.to_vec()));
+
+        let Some((version, bcs)) = self
+            .0
+            .run_query_async(move |conn| query.get_result::<(i64, Vec<u8>)>(conn).optional())
+            .await?
+        else {
+            return Err(Error::PackageNotFound(id));
+        };
+
+        let version = SequenceNumber::from_u64(version as u64);
+        let object = bcs::from_bytes::<Object>(&bcs)?;
+        make_package(id, version, &object)
+    }
+
+    async fn update(&self, _object: &Object) -> Result<()> {
+        unimplemented!("Package update is not implemented")
+    }
+}
+
+#[derive(DBMapUtils)]
+pub struct PackageStoreTables {
+    pub(crate) packages: DBMap<ObjectID, Object>,
+}
+
+impl PackageStoreTables {
+    pub fn new(path: &Path) -> Arc<Self> {
+        Arc::new(Self::open_tables_read_write(
+            path.to_path_buf(),
+            MetricConf::default(),
+            None,
+            None,
+        ))
+    }
+    pub(crate) fn update(&self, package: &Object) -> Result<()> {
+        let mut batch = self.packages.batch();
+        batch
+            .insert_batch(&self.packages, std::iter::once((package.id(), package)))
+            .map_err(Error::TypedStore)?;
+        batch.write().map_err(Error::TypedStore)?;
+        Ok(())
+    }
+}
+
+/// Store which keeps package objects in a local rocksdb store. It is expected that this store is
+/// kept updated with latest version of package objects while iterating over checkpoints. If the
+/// local db is missing (or gets deleted), packages are fetched from a full node and local store is
+/// updated
+pub struct LocalDBPackageStore {
```

Comment:
This struct implements `PackageStore` trait by keeping a local rocksdb store of package objects. DB is kept updated with latest package versions while iterating over checkpoints in pipelines